### ISSUE TITLE
bids-matlab: a "wish-list" of enhancements and requests for new features

### ISSUE THREAD

**Issue by Remi-Gau** [bids-standard/bids-matlab#60]
## List what you would need from BIDS-matlab

As discussed in the last meeting (see #59) we will be keeping track of what people would like bids-matlab to do in this issue.

I will keep this top message updated with new ideas from the discussion below.

---

- a code base that keeps up with new BIDS extensions
- multimodal, not only MRI

---
### Moved to their own issue

- using a schema (see #76) 
> I want the data structures to be formally described (in json/yaml) and not in the MATLAB code, that representation should be shared with pybids (MATLAB has support for JSON - see [here](https://fr.mathworks.com/help/mps/restfuljson/json-representation-of-matlab-data-types.html)

- document current features (see #77)
> documentation is missing: what can the code actually do and not do 

- help with reading BIDS model (see #46) 

- add basic "write" functions (see #47)
> To help with bids compliant filename generation and the creation of json files and data dictionary 

- read and parse derivatives  (see #50)
> Like fMRIprep output

- Initialize derivative data set with a copy of the raw data (see #78)

- Change handling of multi-session dataset (see #63) 

- better handling of the metadata (see #79)
> reading and representing the metadata and the structure, I know how to deal with the data itself

**Comment by Remi-Gau** [comment]
>     * I want the data structures to be formally described (in json/yaml) and not in the MATLAB code, that representation should be shared with pybids (MATLAB has support for JSON - see [here](https://fr.mathworks.com/help/mps/restfuljson/json-representation-of-matlab-data-types.html)

I wonder if we can use the [BIDS schema](https://github.com/bids-standard/bids-specification/tree/master/src/schema) that is being developed and used by the validator to help with this.

**Comment by apjanke** [comment]
Hi! Sorry if I'm butting in here, since this isn't my project. I'm curious about this:

> I want the data structures to be formally described (in json/yaml) and not in the MATLAB code,

Could you go in to this a little bit more, particularly with respect to how you see the BIDS data structures represented as Matlab datatypes inside running Matlab code? Matlab's `jsondecode()` decodes JSON data to generic trees of Matlab structs, cells, strings and numerics. Are you saying that's how you want the data structures to be represented throughout the lifetime of execution?

To my mind, it might be nice to have custom Matlab `classdef` objects corresponding to the BIDS schema structures, even if the "canonical representation" or "source of truth" is the JSON schema. That way you could have eager data validation, custom display formats, and methods attached to the data structures at various levels. (Might be worthwhile for tab-completion support, if nothing else.)

Maybe there'd be a use case for something like the Java XML Beans binding approach, where it maps an XML schema to Java object sets - you could map the BIDS JSON Schema entities to Matlab classes, but still allow for property constraints, other validation, and custom methods added to them?

**Comment by ChristophePhillips** [comment]
There are a few circumstances (see here under) where I would like to initialize the `derivatives` with a copy, complete or partial, of the original full BIDS dataset. I have been working on a `prepare_derivatives.m` for that specific purpose with some features, such as:
- copy the data of all or a sub-list of subjects;
- if there are multiple sessions, copy the data from all or only some sessions;
- copy all or a sub-list of modalities, and for `func` data possibly all or some tasks;
- unzip the `.nii.gz` files and 4D-to-3D unpacking the `func` and `dwi` images, if requested;
- decide where that `derivatives/tool` subfolder will be created, i.e. not necessarily next to the "raw" data in the BIDS folders.

Why would I want to do that? Well because, it is convenient to work on a subset of data, for example
- to develop a processing pipeline with 5 subjects instead of 100 or simply to keep the subjects that are kept for the specific analysis, i.e. leave out some "outliers"
- analysis can focus on a single session for some question and therefore the processing will be different from that combing multiple sessions
- analysis can also focus on a single modality (or task for `func` data), therefore no need to bring along everything
- some operations in SPM's standard special processing pipelines, e.g. realignment and coregistration, do change the header of the images, so they ought to be in a `derivatives` subfolder before any processing. Moreover SPM and some other tools, would prefer dealing with a series of 3D images rather than a single 4D volume, e.g. ditching the 1st few functional images into a "dummy" subfolder.
- if the dataset is saved on an external storage place (like the centralized "mass-storage" system at ULiège), one cannot directly work on this server and therefore data need to be copied locally before any processing is applied, i.e. I might as well call this `derivatives\some_tool_or_step`.

**Comment by Remi-Gau** [comment]
> There are a few circumstances (see here under) where I would like to initialize the `derivatives` with a copy, complete or partial, of the original full BIDS dataset. I have been working on a `prepare_derivatives.m` for that specific purpose with some features, 

I have also something similar in our lab pipeline though yours seem to have more features. I also know that spmup by @CPernet has something that does some of that.

So that seems like one the obvious low hanging fruit !!

**Comment by Remi-Gau** [comment]
hey @apjanke 

responding quickly here eventhough I am not the one who suggested this.

> Maybe there'd be a use case for something like the Java XML Beans binding approach, where it maps an XML schema to Java object sets - you could map the BIDS JSON Schema entities to Matlab classes, but still allow for property constraints, other validation, and custom methods added to them?

I am not sure of exactly in which direction this could go in terms of implementation but this could prove very useful if it can prevent us having to write tons of code for generating filenames for each use case in th BIDS world or to help with some of the other "write" aspect that bids-matlab could have.

Not sure we want to do TOO much validation on our side though and I would relegate that to the bids-validator front.

Noob question: how easy would it be to convert the current BIDS yml schema into something that we can use?

**Comment by bmdeen** [comment]
A few basic functions that I would find useful, from the perspective of creating pipelines that generate/process derivatives in a BIDS-compliant manner:

- A function for changing a BIDS filename, based on changes to specific key/value pairs (i.e., changing, adding, and removing key/value pairs to a name, and maintaining the proper order of fields in the name).
- A function for changing one or several fields in a .json metadata file (e.g., to generate a new metadata file after applying some transformation / preproc step to a dataset)
- A non-private version of bids-matlab's get_metadata function. I frequently find myself wanting a simple function to check a specific metadata field of a single file, without having to run both bids.layout and bids.query. I understand the rationale for keeping get_metadata private, but from my perspective it would be very useful to have a public version of something like this command.

**Comment by gllmflndn** [comment]
@ChristophePhillips: `spm_copy` and `spm_mkdir` are tools to help you for what you want to do:
https://en.wikibooks.org/wiki/SPM/BIDS#Formatting_datasets_into_BIDS
that we used here:
https://github.com/spm/MultimodalScripts/blob/master/code/scripted/master_script.m#L49-L76

@Remi-Gau: using a schema is number one on my wish-list: with it, most of the code in +bids/layout.m could be removed and made generic. [This](https://github.com/bids-standard/bids-specification/tree/master/src/schema) is probably all we need - bit sad that it is using yet another format unnecessarily but it's great it's now available!

**Comment by Remi-Gau** [comment]
> @ChristophePhillips: `spm_copy` and `spm_mkdir` are tools to help you for what you want to do:
> https://en.wikibooks.org/wiki/SPM/BIDS#Formatting_datasets_into_BIDS
> that we used here:
> https://github.com/spm/MultimodalScripts/blob/master/code/scripted/master_script.m#L49-L76

Actually had opened an [issue](https://github.com/cpp-lln-lab/CPP_BIDS/issues/58) on one of our repo to get `spm_mkdir` and `spm_copy` out of spm. So I would :heart: to have them as part of bids matlab. 

One headache to keep in mind: datasets curated with datalad have their content stored with git annex (I need to finish a PR about that on the [datalad handbook](https://github.com/datalad-handbook/book/pull/570)). So a simple call to `copyfile` will not follow the symbolic link and you just end up with just a bunch of broken links. 

2 options: 
- the user must make sure they have run `datalad unlock` the files to copy 
- try a system call to `cp -L` and catch with `copyfile` if it fails (this is the hacky way of doing things we are currentty using: see [here](https://github.com/Remi-Gau/CPP_BIDS_SPM_pipeline/blob/bf850e944b0f9e2ee8381c11b2a3ef21960da54c/src/workflows/bidsCopyRawFolder.m#L97)

The second option is going to make Windows users cry though... If they use datalad: no a huge user base at the moment but that too could grow.

Any other ways around this that would make everyone happy?

**Comment by Remi-Gau** [comment]
> @Remi-Gau: using a schema is number one on my wish-list: with it, most of the code in +bids/layout.m could be removed and made generic. [This](https://github.com/bids-standard/bids-specification/tree/master/src/schema) is probably all we need - bit sad that it is using yet another format unnecessarily but it's great it's now available!

I think that if we could automate the conversion of the yml based schema into something suits our need better. That would indeed make the code base a LOT lighter. :-)

**Comment by gllmflndn** [comment]
> I think that if we could automate the conversion of the yml based schema into something suits our need better. That would indeed make the code base a LOT lighter. :-)

If we want to use this schema (which I think we should), we can either 1/ use or implement a (simplified) YAML parser or 2/ convert it to JSON first and use this in bids-matlab. I would recommend option 2 and see whether the JSON-converted schema could be stored in the bids-specification repository too (the YAML version being the authoritative one and the JSON one being automatically converted and kept in sync).

**Comment by gllmflndn** [comment]
> Actually had opened an [issue](https://github.com/cpp-lln-lab/CPP_BIDS/issues/58) on one of our repo to get `spm_mkdir` and `spm_copy` out of spm. So I would ❤️ to have them as part of bids matlab.

https://github.com/cpp-lln-lab/CPP_BIDS/issues/58#issuecomment-665799330 :disappointed_relieved:

> One headache to keep in mind: datasets curated with datalad have their content stored with git annex (I need to finish a PR about that on the [datalad handbook](https://github.com/datalad-handbook/book/pull/570)). So a simple call to `copyfile` will not follow the symbolic link and you just end up with just a bunch of broken links.

I think we should make a system() call only out of necessity. We could test for symlinks within a `isunix` condition and only use `cp -L` for these?

**Comment by Remi-Gau** [comment]
> > Actually had opened an [issue](https://github.com/cpp-lln-lab/CPP_BIDS/issues/58) on one of our repo to get `spm_mkdir` and `spm_copy` out of spm. So I would heart to have them as part of bids matlab.
> 
> [cpp-lln-lab/CPP_BIDS#58 (comment)](https://github.com/cpp-lln-lab/CPP_BIDS/issues/58#issuecomment-665799330) disappointed_relieved

hahaha 
Nothing personal against SPM. Just wanted to keep that repo as standalonish as possible. :-) If people want to use it to test with psychtoolbox, I would not want to them to install SPM just for that. :-p


> > One headache to keep in mind: datasets curated with datalad have their content stored with git annex (I need to finish a PR about that on the [datalad handbook](https://github.com/datalad-handbook/book/pull/570)). So a simple call to `copyfile` will not follow the symbolic link and you just end up with just a bunch of broken links.
> 
> I think we should make a system() call only out of necessity. We could test for symlinks within a `isunix` condition and only use `cp -L` for these?

Yes that sounds like a good idea to me: my hack was very much tailored to my current needs which is very UNIX + datalad heavy. :-)

**Comment by Remi-Gau** [comment]
> If we want to use this schema (which I think we should), we can either 1/ use or implement a (simplified) YAML parser or 2/ convert it to JSON first and use this in bids-matlab. I would recommend option 2 and see whether the JSON-converted schema could be stored in the bids-specification repository too (the YAML version being the authoritative one and the JSON one being automatically converted and kept in sync).

I prefer option 2 as well. I think that this automatic conversion could be part of the infrastructure of the specs. Would need to raise the issue in the specs repo and sor that there before we can start working on this.

**Comment by nbeliy** [comment]
### Generalize bids layout function

In actual state `layout` function checks for comliency with current (hardcoded) bids schema. Which will automatically made it unusable if dataset do not follow defined schema for whatever reason. For example if one uses an old bidsified dataset, not compatible with actual state.

So I propose to move the checks for modalities, file extentions, entities to `validate` function, and make `layout` working on any dataset which follows bids syntax: `sub-<>_(<key>-<value>)*_<suffix>.<extention>`. 

It will not only simplify the code (all complex ifs will move to `validate`), but also make core code independent of version of BIDS, simplifying the maintenance. 

Additionally, if derivatives follow the same naming rules as original bids, the `layout` will also work on it, facilitating the data treatment.

### Make query return subset of layout instead of flat list

In current state `querry` returns  flat lists of found data, subjects etc. This may create some difficulties during data processing, for ex. if one need to access to metadata in json file, he need to reparse the filenames.

I propose that `querry` returns instead the either subset subject field, or full layout containing only selected data files. Retrieve a flat list of files from it is trivial, so it will nod add complexity if one just need data files, but it will provide additional metadata together with data files, if they are needed.

### In layout provide links to files needed to analyse given image

In some modalities (meg/eeg. func) images are accompanied by tabular data, like `events.tsv`. In actual state, one need to extract data file names and then replace the suffix to access these tabular files. I propose instead introduce a new field `dependancies` which will contain the paths to corresponding tabular files. The fmap images can also be included in it.
This way, it will be which files are needed to analyse each given image.

**Comment by Remi-Gau** [comment]
@nbeliy Actually moving your comment to its own issue to help with the discussion.

**Comment by nbeliy** [comment]
### Add prefixes to parsed derivative file names 

In the SPM, the output files are usually created with a prefix to original file name. For usual names it is an easy way for ordinary files, but it will break the bidsified names, as nothing can go before `sub-`. 
Thus, if user want to daisy-chain processing step maintaining he need rename files after each processing step. 
However, if `bids.layout` can support prefix (SPM allow the setting of prefix, so it can be separated from sub by `_`), then the output of each process can be readded directly by `layout`. Obviously this prefix should be defined only for derivatives.

**Comment by Remi-Gau** [comment]
> ### Add prefixes to parsed derivative file names
> 
> In the SPM, the output files are usually created with a prefix to original file name. For usual names it is an easy way for ordinary files, but it will break the bidsified names, as nothing can go before `sub-`.
> Thus, if user want to daisy-chain processing step maintaining he need rename files after each processing step.
> However, if `bids.layout` can support prefix (SPM allow the setting of prefix, so it can be separated from sub by `_`), then the output of each process can be readded directly by `layout`. Obviously this prefix should be defined only for derivatives.

Ideally for SPM I would recommend using the dependency options of its batch system to link the different analysis step  (also because it is an embedded way to create a log file that tracks provenance) but I will try add it to the list of options.

**Comment by Remi-Gau** [comment]
Will close this issue and have every feature request be handled with a different issue / discussions.

To get an overview we will use the github kanban boards.
